content/en/blog/_posts/2022-04-19-csi-nodeexpandsecret.md

```diff
@@ -0,0 +1,183 @@
+Even-though Kubernetes CSI have implemented similar mechanism for Controller
+side operations, ie secretRef field available in the csi PV source and making
+use of it while controllerExpand request has been sent to the CSI driver, similar
+field was missing in the nodeExpandVolume request.  This was causing some trouble
+for different CSI drivers to fulfill below scenarios while it perform NodeExpansion
+operation.
+
+- At times, the CSI driver need to check the actual size of the backend volume/image
+  before proceeding on node side FS resize operation to avoid false positive returns
+  from the  backend storage cluster on fs resize operation.
+
+- Encrypted device with LUKs, which need the passphrase in order to resize
+  the device on the node.
+
+- For various validations at time of node expansion the CSI driver has to be connected
+  to the backend storage cluster, if the secretRef is part of the
+  nodeExpandVolume request the CSI driver can make use of the same and connect to
+  the storage cluster  to perform the cluster operations.
+
+## How does it work?
+
+To enable this functionality from this version of Kubernetes, we have introduced
+a new feature gate called CSINodeExpandSecret. Once the feature gate is enabled
+in the cluster, the secretRef field will be added to the NodeExpandVolume request
+which has been sent from the kubernetes orchestrator to the CSI driver.
+The admin can specify these secrets as an opaque parameter in storage class as
+he/she specifies other CSI secrets today.  The secrets mention in the storage class
+has to carry the key name as shown below
+
+ `csi.storage.k8s.io/node-expand-secret-name: test-secret
+  csi.storage.k8s.io/node-expand-secret-namespace: default`
+
+If feature gates are enabled and storage class carries the above secret configuration,
+the subject CSI provisioner will receive the secrets as part of the NodeExpansion request.
+
+CSI volumes that require secrets for online expansion will have NodeExpandSecretRef
+field set. If not set, the NodeExpandVolume CSI RPC call will be made without a secret.
+
+
+
+## Trying it out
+
+
+Step 1: enable CSINodeExpandSecret feature gates ( Refer featuregate
+documentation (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/))
+
+Step 2: Secret and storage class creation
+
+Ex: secret.yaml:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: test-secret
+data:
+  username: bXktYXBw
+  password: Mzk1MjgkdmRnN0pi
+```
+
+Ex: sc.yaml:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-hostpath-sc
+parameters:
+  csi.storage.k8s.io/node-expand-secret-name: test-secret
+  csi.storage.k8s.io/node-expand-secret-namespace: default
+provisioner: hostpath.csi.k8s.io
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
+allowVolumeExpansion: true
+```
+
+
+## Example output:
+
+If the pvc was created successfully, we can see it in the pv.spec.csi with cmd `kubectl get pv <pv_name> -o yaml`
```

Comment:
```suggestion
If the `pvc` was created successfully, we can see it in the `pv.spec.csi` with command `kubectl get pv <pv_name> -o yaml`
```

Reply:
This needs a bigger rewrite:
- avoid using “we”
- don't start a new 2nd-level section half way through a demo
- make it clear what the reader has to do do get to a point where they can create a working PersistentVolumeClaim
- use long names, eg persistentvolumeclaim not pvc, in example commands.